code/centauri/ibc/modules/ics02_client.md

```diff
@@ -0,0 +1,365 @@
+- `ClientMisbehaviour` -  A `MsgSubmitMisbehaviour` has been processed and the client has been frozen.
+
+The client events are defined [`here`](/code/centauri/ibc/modules/src/core/ics02_client/events.rs)
+
+### Upgrading a Client
+
+A client upgrade is required when there is a breaking change in a chain's light client protocol.  
+
+To upgrade a client, a client upgrade path for both the client state and consensus state must have been predefined in either the client state or as a constant defined when constructing the light client
+
+The chain undergoing the upgrade should then commit the upgraded client and consensus states to its storage using the expected upgrade paths.  
+
+The `MsgUpgradeClient` can now be submitted with the proof for the upgrade.
+
+### Supporting Multiple Light Client Types
+
+The code below describes how to allow support for multiple light client types using the macros provided in `ibc-derive` crate.
+
+```rust
+#[derive(Clone, Debug, PartialEq, Eq, ibc_derive::ClientDef)]
+pub enum AnyClient {
+	Grandpa(ics10_grandpa::client_def::GrandpaClient),
+	Beefy(ics11_beefy::client_def::BeefyClient),
+	Tendermint(ics07_tendermint::client_def::TendermintClient),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum AnyUpgradeOptions {
+	Grandpa(ics10_grandpa::client_state::UpgradeOptions),
+	Beefy(ics11_beefy::client_state::UpgradeOptions),
+	Tendermint(ics07_tendermint::client_state::UpgradeOptions),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, ibc_derive::ClientState, ibc_derive::Protobuf)]
+pub enum AnyClientState {
+	#[ibc(proto_url = "GRANDPA_CLIENT_STATE_TYPE_URL")]
+	Grandpa(ics10_grandpa::client_state::ClientState<HostFunctionsManager>),
+	#[ibc(proto_url = "BEEFY_CLIENT_STATE_TYPE_URL")]
+	Beefy(ics11_beefy::client_state::ClientState<HostFunctionsManager>),
+	#[ibc(proto_url = "TENDERMINT_CLIENT_STATE_TYPE_URL")]
+	Tendermint(ics07_tendermint::client_state::ClientState<HostFunctionsManager>),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, ibc_derive::ConsensusState, ibc_derive::Protobuf)]
+pub enum AnyConsensusState {
+	#[ibc(proto_url = "GRANDPA_CONSENSUS_STATE_TYPE_URL")]
+	Grandpa(ics10_grandpa::consensus_state::ConsensusState),
+	#[ibc(proto_url = "BEEFY_CONSENSUS_STATE_TYPE_URL")]
+	Beefy(ics11_beefy::consensus_state::ConsensusState),
+	#[ibc(proto_url = "TENDERMINT_CONSENSUS_STATE_TYPE_URL")]
+	Tendermint(ics07_tendermint::consensus_state::ConsensusState),
+}
+
+#[derive(Clone, Debug, ibc_derive::ClientMessage)]
+#[allow(clippy::large_enum_variant)]
+pub enum AnyClientMessage {
+	#[ibc(proto_url = "GRANDPA_CLIENT_MESSAGE_TYPE_URL")]
+	Grandpa(ics10_grandpa::client_message::ClientMessage),
+	#[ibc(proto_url = "BEEFY_CLIENT_MESSAGE_TYPE_URL")]
+	Beefy(ics11_beefy::client_message::ClientMessage),
+	#[ibc(proto_url = "TENDERMINT_CLIENT_MESSAGE_TYPE_URL")]
+	Tendermint(ics07_tendermint::client_message::ClientMessage),
+}
+
+impl Protobuf<Any> for AnyClientMessage {}
+
+impl TryFrom<Any> for AnyClientMessage {
```

Comment:
do we need to show this impl?

Reply:
I think I'll just summarise it in text format.